crates/codegen/language/definition/src/compiler/versions.rs

```diff
@@ -0,0 +1,215 @@
+use semver::Version;
+use std::{
+    cmp::{max, min},
+    fmt::Display,
+    iter::once,
+};
+
+const MAX_VERSION: Version = Version::new(u64::MAX, u64::MAX, u64::MAX);
+
+#[derive(Clone, Debug)]
+pub struct VersionRange {
+    pub inclusive_start: Version,
+    pub exclusive_end: Version,
+}
+
+impl VersionRange {
+    pub fn starting_from(inclusive_start: &Version) -> Self {
+        return Self {
+            inclusive_start: inclusive_start.to_owned(),
+            exclusive_end: MAX_VERSION,
+        };
+    }
+
+    pub fn between(inclusive_start: &Version, exclusive_end: &Version) -> Self {
+        if inclusive_start <= exclusive_end {
+            return Self {
+                inclusive_start: inclusive_start.to_owned(),
+                exclusive_end: exclusive_end.to_owned(),
+            };
+        } else {
+            // This is an invalid state, which we still produce errors for.
+            // However, for validation to continue, let's correct the order:
+            return Self {
+                inclusive_start: exclusive_end.to_owned(),
+                exclusive_end: inclusive_start.to_owned(),
+            };
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        return self.inclusive_start == self.exclusive_end;
+    }
+}
+
+impl Display for VersionRange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.inclusive_start == MAX_VERSION {
+            write!(f, "MAX")?;
+        } else {
+            write!(f, "{}", self.inclusive_start)?;
+        }
+
+        write!(f, "..")?;
+
+        if self.exclusive_end == MAX_VERSION {
+            write!(f, "MAX")?;
+        } else {
+            write!(f, "{}", self.exclusive_end)?;
+        }
+
+        return Ok(());
+    }
+}
+
+#[derive(Debug)]
+pub struct VersionSet {
+    ranges: Vec<VersionRange>,
+}
+
+impl VersionSet {
+    pub fn empty() -> Self {
+        return Self { ranges: vec![] };
+    }
+
+    pub fn single(range: VersionRange) -> Self {
+        let mut instance = Self::empty();
+        instance.add(&range);
+        return instance;
+    }
+
+    pub fn union(ranges: impl IntoIterator<Item = VersionRange>) -> Self {
+        let mut instance = Self::empty();
+
+        for range in ranges.into_iter() {
+            instance.add(&range);
+        }
+
+        return instance;
+    }
+
+    pub fn is_empty(&self) -> bool {
+        return self.ranges.is_empty();
+    }
+
+    pub fn add(&mut self, range: &VersionRange) {
```

Comment:
Is there a reason why this is mutable but the difference is immutable but creates a new value?

Reply:
Missed that from an earlier refactoring. Good catch. Will fix it here.

Reply:
Not always. Otherwise, this serves as an accumulator.

Reply:
Left from an earlier refactoring. Good catch. Fixed.

Reply:
Not always. It sometimes serves as an accumulator, instead of cloning `self` when it is not changed.